Link Yosemite to NotificationExtension which depends on it

mokagio opened the pull request:
## Description

Today, @itsmeichigo reported a CI failure p1726195899322149-slack-CC7L49W13 and identified #13896 as the first PR that, once merged, resulted in [the failure](https://buildkite.com/automattic/woocommerce-ios/builds/24609#_).

The failure:

```
ld: warning: Could not find or use auto-linked framework 'Yosemite': framework 'Yosemite' not found
❌ Undefined symbols for architecture arm64
❌   "protocol witness table for Yosemite.NotificationAction : Yosemite.Action in Yosemite", referenced from:
❌   "nominal type descriptor for Yosemite.NotificationAction", referenced from:
❌   "type metadata for Yosemite.NotificationAction", referenced from:
❌ ld: symbol(s) not found for architecture arm64
❌ clang: error: linker command failed with exit code 1 (use -v to see invocation)

[...]

** TEST BUILD FAILED **
The following build commands failed:
	Ld /opt/ci/builds/builder/automattic/woocommerce-ios/DerivedData/Build/Products/Debug-iphonesimulator/NotificationExtension.appex/NotificationExtension normal (in target 'NotificationExtension' from project 'WooCommerce')
```

After a bit of trial and error, made slow by the fact that I am not able to reproduce the issue locally (I tried all the usual things, Clean, removed `DerivedData`, run locally the same command that CI runs, ...), I noticed the following.

The changes in 1fd27830a602886f96f5bfe2b6c346e851880041 and 2e7b7e063c4d8e5446dc7d5a6962f1405ff6d2a0 add `import Yosemite` to `MarkOrderAsReadUseCase.swift`. That file is part of various targets, namely WooCommerce, WooCommerce Watch App, and NotificationExtension:

<img width="270" alt="image" src="https://github.com/user-attachments/assets/f967285b-5951-4ccc-a6c6-22396a0abf90">
(_Why is it in the test target instead of being accessible there via `@testable import`?_)

WooCommerce links to Yosemite. The Watch target does not, but it's not a problem because of the `#if canImport(Yosemite)` check. But NotificationExtension did not link to Yosemite, could it then be that the `#if canImport(Yosemite)` check passed (it uses the platform destinations?) but then the build failed because the framework is not in the target?

69da800d1c4d9cbfe1c077a1720f70c99d2270a9 adds Yosemite to NotificationExtension and the build passes, which gives credence to the my guess.

The only thing that I cannot explain is why a local clean build does not run in the problem and why CI in #13896 passed and allowed the merge. My best guess is that this has something to do with Xcode parallelizing builds in a non deterministic order resulting in the framework being available during compilation (even if not directly imported) locally or in some builds but not in other. I don't know enough about the guts of the builds system to say whether this guess is solid or not.

### Where to go from here

This issue was caused by the anti pattern of sharing logic between targets by including the logic source in all targets as opposed to isolating the logic in a framework. If the logic had been isolated in a framework, the Yosemite dependency would have had to be addressed at the framework level and, once sorted, all the consumers would have access to the appropriate code without additional setup (_I think_).

I don't know the context that resulted in this design and I assume there was a reasonable tradeoff at the time to go down that path. Nevertheless, going forward I would recommend moving the this kind of shared logic in a framework or Swift package.

_If that doesn't happen, hopefully this verbose description of mine will help future versions of ourselves fix this kind of issue faster_ 😅 

For what is worth, this is something WordPress iOS was also suffering from the last time I checked...

## Steps to reproduce
N.A.

## Testing information
Since this was a CI failure, I green CI in this build should suffixe

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

mokagio committed:
Remove dead "Vendors" reference in project tree

mokagio committed:
Use named block parameter in `build_for_testing` lane

mokagio committed:
Remove comment that didn't add any information

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

mokagio committed:
Bump release-toolkit to latest version 12.1.0

Just so it stops asking me when I run Fastlane lanes

mokagio committed:
Move header doc for method after #if conditional for definition

mokagio committed:
HACK – Disable code to verify if it is source of failure in CI

mokagio committed:
Revert "HACK – Disable code to verify if it is source of failure in CI"

This reverts commit 469c26868a119f78e85f228d91b4a08a0ef4a1f8.

mokagio committed:
Access `NotificationAction` without `Yosemite` namespace prefix

mokagio commented on WooCommerce/Classes/Model/MarkOrderAsReadUseCase.swift:
```diff
@@ -18,15 +18,16 @@ struct MarkOrderAsReadUseCase {
+
+#if canImport(Yosemite)
     /// Async method that marks the order note as read if it is the notification for the last order.
     /// We do it in a way that first we syncronize notification to get the remote `Note`
     /// and then we compare local `orderID` with the one from remote `Note`.
     /// If they match we mark it as read.
     /// Returns syncronized note if marking was successful and error if some error happened
-#if canImport(Yosemite)
```
Moved the header doc within the `#if` branch that conditionally defines the method the header doc documents

mokagio commented on WooCommerce/WooCommerce.xcodeproj/project.pbxproj:
```diff
@@ -6240,13 +6240,6 @@
-			);
-			path = Vendors;
-			sourceTree = "<group>";
-		};
```
Dead references to an empty group.

mokagio commented on fastlane/Fastfile:
```diff
@@ -782,25 +782,17 @@ platform :ios do
-  # -----------------------------------------------------------------------------------
-  # Usage:
-  # bundle exec fastlane build_for_testing
-  #####################################################################################
```
This comment didn't tell us anything that the lane name didn't already say.

mokagio committed:
Remove duplicated `WooConstants.swift` source in Watch App target

mokagio committed:
Link Yosemite to NotificationExtension target

The changes in 1fd27830a602886f96f5bfe2b6c346e851880041 and
2e7b7e063c4d8e5446dc7d5a6962f1405ff6d2a0 add `import Yosemite` to
`MarkOrderAsReadUseCase.swift`.

That file is part of WooCommerce, WooCommerce Watch App, and
NotificationExtension.

WooCommerce links to Yosemite. The Watch target does not, but it's not a
problem because of the `#if canImport(Yosemite)` check.

But what about NotificationExtension? It did not have Yosemite before
this change. Might it be the reason for the CI failure:

```
ld: warning: Could not find or use auto-linked framework 'Yosemite': framework 'Yosemite' not found
❌ Undefined symbols for architecture arm64
❌   "protocol witness table for Yosemite.NotificationAction : Yosemite.Action in Yosemite", referenced from:
❌   "nominal type descriptor for Yosemite.NotificationAction", referenced from:
❌   "type metadata for Yosemite.NotificationAction", referenced from:
❌ ld: symbol(s) not found for architecture arm64
❌ clang: error: linker command failed with exit code 1 (use -v to see invocation)

[...]

The following build commands failed:
	Ld /opt/ci/builds/builder/automattic/woocommerce-ios/DerivedData/Build/Products/Debug-iphonesimulator/NotificationExtension.appex/NotificationExtension normal (in target 'NotificationExtension' from project 'WooCommerce')
```

See for example
https://buildkite.com/automattic/woocommerce-ios/builds/24623#0191e996-5541-4a94-9454-30f771bae463/7218-13610

mokagio commented on Gemfile.lock:
```diff
@@ -207,7 +208,7 @@ GEM
     fastlane-plugin-appcenter (2.1.2)
     fastlane-plugin-sentry (1.24.0)
       os (~> 1.1, >= 1.1.4)
-    fastlane-plugin-wpmreleasetoolkit (12.0.0)
+    fastlane-plugin-wpmreleasetoolkit (12.1.0)
```
This is unrelated to the fix. It was just handy to upgrade to:

1. Make this available for the next code freeze
2. Not get prompted every time I run the tests locally

mokagio commented on WooCommerce/Classes/Model/MarkOrderAsReadUseCase.swift:
```diff
@@ -18,15 +18,16 @@ struct MarkOrderAsReadUseCase {
-#if canImport(Yosemite)
     static func markOrderNoteAsReadIfNeeded(stores: StoresManager, noteID: Int64, orderID: Int) async -> Result<Note, Error> {
         let syncronizedNoteResult: Result<Note, Error> = await withCheckedContinuation { continuation in
-            let action = Yosemite.NotificationAction.synchronizeNotification(noteID: noteID) { syncronizedNote, error in
+            let action = NotificationAction.synchronizeNotification(noteID: noteID) { syncronizedNote, error in
```
I thought maybe there was something odd with the namespace access that resulted in the build failure so I tried removing it. But that was not the reason.

If you find the code clear with the prefix, feel free to restore it.

mokagio commented on WooCommerce/WooCommerce.xcodeproj/project.pbxproj:
```diff
@@ -1017,7 +1017,6 @@
 		269FFA472BF40768004E6B86 /* WooFoundationWatchOS.framework in Embed Frameworks */ = {isa = PBXBuildFile; fileRef = 269FFA452BF40768004E6B86 /* WooFoundationWatchOS.framework */; settings = {ATTRIBUTES = (CodeSignOnCopy, RemoveHeadersOnCopy, ); }; };
 		269FFA482BF516BA004E6B86 /* Logging.swift in Sources */ = {isa = PBXBuildFile; fileRef = 933A27362222354600C2143A /* Logging.swift */; };
 		269FFA4B2BF544C9004E6B86 /* MyStoreViewModel.swift in Sources */ = {isa = PBXBuildFile; fileRef = 269FFA4A2BF544C9004E6B86 /* MyStoreViewModel.swift */; };
-		26A000772BE9D9030071DB1E /* WooConstants.swift in Sources */ = {isa = PBXBuildFile; fileRef = B5D1AFBF20BC67C200DB0E8C /* WooConstants.swift */; };
```
This was a duplicated source in WooFoundation. I noticed the warning in the Xcode navigator.

mokagio commented on WooCommerce/WooCommerce.xcodeproj/project.pbxproj:
```diff
@@ -6010,6 +6011,7 @@
 				2608370B2AA66E4B0004A12B /* UserNotificationsUI.framework in Frameworks */,
 				260837092AA66E4B0004A12B /* UserNotifications.framework in Frameworks */,
 				048074D723A784C0C998FCB8 /* Pods_NotificationExtension.framework in Frameworks */,
+				3FA96DF42C94043200CDA78F /* Yosemite.framework in Frameworks */,
```
These new `Yosemite.framework` entries are the actual fix.

mokagio commented on fastlane/Fastfile:
```diff
@@ -782,25 +782,17 @@ platform :ios do
       build_for_testing: true,
-      device: options[:device],
-      deployment_target_version: options[:ios_version]
+      device: device,
+      deployment_target_version: ios_version
```
Converted to the new preferred named block parameters syntax while I was at it.
